src/TypeInference/GenerateConstraints/Kinds.hs

```diff
@@ -51,6 +55,259 @@ newKVar = do
+removeKVars (TST.TyInter loc knd ty1 ty2) = return (TST.TyInter loc knd ty1 ty2) -- TODO -- ty1 and ty2 may contain kvars
+removeKVars tr@TST.TyRec{}                = return tr                            -- TODO -- recursive type and recvar may contain kvars
+-- primitive types don't contain a type
+removeKVars i64@TST.TyI64{}               = return i64
+removeKVars f64@TST.TyF64{}               = return f64
+removeKVars ch@TST.TyChar{}               = return ch
+removeKVars str@TST.TyString{}            = return str
+-- only the contained type needs to be checked
+removeKVars (TST.TyFlipPol pol ty)        = do 
+  ty' <- removeKVars ty
+  return $ TST.TyFlipPol pol ty'
+
+
+annotTy :: RST.Typ pol -> DataDeclM (TST.Typ pol)
+annotTy (RST.TySkolemVar loc pol tv) = do 
+  polyknd <- gets declKind
+  return $ TST.TySkolemVar loc pol (getKindSkolem polyknd tv) tv 
+-- uni vars should not appear in data declarations
+annotTy (RST.TyUniVar loc _ _) = throwOtherError loc ["UniVar should not appear in data declaration"]
+annotTy (RST.TyRecVar loc pol tv) = do
+  rvs <- gets boundRecVars 
+  case M.lookup tv rvs of 
+    Nothing -> throwOtherError loc ["Unbound RecVar " <> ppPrint tv <> " in Xtor"]
+    Just knd -> return $ TST.TyRecVar loc pol knd tv
+annotTy (RST.TyData loc pol xtors) = do 
+  let xtnms = map RST.sig_name xtors
+  xtorKinds <- mapM lookupXtorKind xtnms
+  let allEq = compXtorKinds xtorKinds  
+  case allEq of 
+    Nothing -> throwOtherError loc ["Not all xtors have the same return kind"]
+    Just mk -> do 
+      xtors' <- mapM (\x -> lookupXtorSig loc x pol) xtnms
+      return $ TST.TyData loc pol mk xtors' 
+  where 
+    compXtorKinds :: [MonoKind] -> Maybe MonoKind
+    compXtorKinds [] = Nothing 
+    compXtorKinds [mk] = Just mk
+    compXtorKinds (xtor1:xtor2:rst) = if xtor1==xtor2 then compXtorKinds (xtor2:rst) else Nothing
+annotTy (RST.TyCodata loc pol xtors) = do 
+  let xtnms = map RST.sig_name xtors
+  xtorKinds <- mapM lookupXtorKind xtnms
+  let allEq = compXtorKinds xtorKinds  
+  case allEq of 
+    Nothing -> throwOtherError loc ["Not all xtors have the same return kind"]
+    Just mk -> do 
+      xtors' <- mapM (\x -> lookupXtorSig loc x (RST.flipPolarityRep pol)) xtnms
+      return $ TST.TyCodata loc pol mk xtors' 
+  where 
+    compXtorKinds :: [MonoKind] -> Maybe MonoKind
+    compXtorKinds [] = Nothing 
+    compXtorKinds [mk] = Just mk
+    compXtorKinds (xtor1:xtor2:rst) = if xtor1==xtor2 then compXtorKinds (xtor2:rst) else Nothing
+annotTy (RST.TyDataRefined loc pol tyn xtors) =  do 
+  xtors' <- mapM annotXtor xtors
+  tyn' <- gets declTyName
+  if tyn == tyn' then do
+    polyknd <- gets declKind
+    return $ TST.TyDataRefined loc pol (CBox $ returnKind polyknd) tyn xtors' 
+  else do 
+    decl <- lookupTypeName loc tyn
+    return $ TST.TyDataRefined loc pol (CBox $ returnKind $ TST.data_kind decl) tyn xtors' 
```

Comment:
ist im Fall von `tyn == tyn'` `gets declKind` und `lookupTypeName loc tyn` nicht das Gleiche?
Wenn es das Gleiche ist, warum dann die Unterscheidung?

Reply:
Weil in diesem Fall tyn gar nicht im Environment steht, das wird ja berechnet, und dann wird der erst ins environment geschrieben